components/script/dom/datatransfer.rs

```diff
@@ -0,0 +1,149 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::cell::Cell;
+
+use dom_struct::dom_struct;
+use js::rust::HandleObject;
+
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::DataTransferBinding::DataTransferMethods;
+use crate::dom::bindings::reflector::{reflect_dom_object_with_proto, DomObject, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::DOMString;
+use crate::dom::datatransferitemlist::DataTransferItemList;
+use crate::dom::element::Element;
+use crate::dom::filelist::FileList;
+use crate::dom::window::Window;
+use crate::script_runtime::CanGc;
+
+#[derive(Clone, Copy, Eq, JSTraceable, MallocSizeOf, PartialEq)]
+pub enum Mode {
+    ReadWrite,
+    ReadOnly,
+    Protected,
+}
+
+#[dom_struct]
+pub struct DataTransfer {
+    reflector_: Reflector,
+    drop_effect: DomRefCell<DOMString>,
+    effect_allowed: DomRefCell<DOMString>,
+    mode: Cell<Mode>,
+    items: Dom<DataTransferItemList>,
+}
+
+impl DataTransfer {
+    pub fn new_inherited(item_list: &DataTransferItemList) -> DataTransfer {
+        DataTransfer {
+            reflector_: Reflector::new(),
+            drop_effect: DomRefCell::new(DOMString::from("none")),
+            effect_allowed: DomRefCell::new(DOMString::from("none")),
+            items: Dom::from_ref(item_list),
+            mode: Cell::new(Mode::ReadWrite),
+        }
+    }
+
+    pub fn new_with_proto(
+        window: &Window,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+    ) -> DomRoot<DataTransfer> {
+        let item_list = DataTransferItemList::new(window, proto, can_gc);
```

Comment:
Passing the `proto` value here is incorrect; it is only a valid prototype for the `DataTransfer` object.